### PR TITLE
test(robot): migrate test_snapshot test case

### PR DESCRIPTION
**Pull request by yangchiu**
#### Which issue(s) this PR fixes:

Issue https://github.com/longhorn/longhorn/issues/8259

#### What this PR does / why we need it:

migrate test_snapshot test case

### CODE REVIEW

File: e2e/tests/regression/test_basic.robot [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+*** Test Cases ***
+Volume Basic Test
+    [Tags]  coretest
+    [Documentation]    Test basic volume operations
+    ...
+    ...                1. Check volume name and parameter
+    ...                2. Create a volume and attach to the current node, then check volume states
+    ...                3. Check soft anti-affinity rule
+    ...                4. Write then read back to check volume data
+    When Create volume wrong_volume-name-1.0
```
Comment (c3y1huang) [inline e2e/tests/regression/test_basic.robot:31]: Please follow the BDD style, adopting the Given, When, Then structure.

https://github.com/longhorn/longhorn/issues/8375

**Review by c3y1huang — DISMISSED** [review]

File: e2e/tests/regression/test_basic.robot (hunk 1 above)
Comment (yangchiu) [reply e2e/tests/regression/test_basic.robot]: Refined. Thank you!

File: e2e/tests/regression/test_basic.robot [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    ...
+    ...                1. Check volume name and parameter
+    ...                2. Create a volume and attach to the current node, then check volume states
+    ...                3. Check soft anti-affinity rule
+    ...                4. Write then read back to check volume data
+    Given Create volume wrong_volume-name-1.0
+    Then No volume created
+
+    Given Create volume wrong_volume-name
+    Then No volume created
```
Comment (c3y1huang) [inline e2e/tests/regression/test_basic.robot:35]: These tests involve testing with different names. Can we leverage the [test template](https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#test-templates)?

File: e2e/tests/regression/test_basic.robot [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    And Wait for volume 0 healthy
+    Then Validate volume 0 replicas anti-affinity
+
+    When Write data to volume 0
+    Then Check volume 0 data is intact
+
+    And Detach volume 0
+    And Delete volume 0
+
+Test Snapshot
```
Comment (c3y1huang) [inline e2e/tests/regression/test_basic.robot:51]: I think we should re-organize the tests to improve on the groupings.
```
- e2e/tests/
   +--- volume
   +--- snapshot
   +--- setting
   +--- etc.
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: e2e/tests/regression/test_basic.robot (hunk 2 above)
Comment (roger-ryao) [reply e2e/tests/regression/test_basic.robot:35]: Hi @c3y1huang 
     According to your suggestion regarding the test template, do you mean like the code below?
```python
*** Keywords ***
Create Volumes With Invalid Value
    Create volume wrong_volume-name-1.0
    Create volume wrong_volume-name
    
Volume should NOT be created    
    No volume created

*** Test Cases ***
Volume Basic Test
    [Tags]  coretest
    [Documentation]    Test basic volume creation
    Given Create Volumes With Invalid Value
    Then Volume should NOT be created 
```

File: e2e/tests/regression/test_basic.robot (hunk 3 above)
Comment (yangchiu) [reply e2e/tests/regression/test_basic.robot:51]: @c3y1huang Are `volume`, `snapshot` folders? Could you suggest the robot filename?

File: e2e/keywords/volume.resource [hunk 4]
```diff
@@ -33,6 +33,10 @@ Attach volume ${volume_id}
     ${volume_name} =    generate_name_with_suffix    volume    ${volume_id}
     attach_volume    ${volume_name}
 
+And Attach volume ${volume_id} in maintenance mode
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/keywords/volume.resource]: Refined. Thank you!

File: e2e/libs/utility/utility.py [hunk 5]
```diff
@@ -100,11 +100,14 @@ def apply_cr_from_yaml(filepath):
 
 def get_cr(group, version, namespace, plural, name):
     api = client.CustomObjectsApi()
-    try:
-        resp = api.get_namespaced_custom_object(group, version, namespace, plural, name)
-        return resp
-    except ApiException as e:
-        logging(f"Getting namespaced custom object error: {e}")
+    retry_count, retry_interval = get_retry_count_and_interval()
+    for i in range(retry_count):
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/libs/utility/utility.py]: Refined. Thank you!

File: e2e/tests/regression/test_basic.robot [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+*** Test Cases ***
+Volume Basic Test
+    [Tags]  coretest
+    [Documentation]    Test basic volume operations
+    ...
+    ...                1. Check volume name and parameter
+    ...                2. Create a volume and attach to the current node, then check volume states
+    ...                3. Check soft anti-affinity rule
+    ...                4. Write then read back to check volume data
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/tests/regression/test_basic.robot]: Refined. Thank you!

File: e2e/tests/regression/test_basic.robot (hunk 2 above)
Comment (yangchiu) [reply e2e/tests/regression/test_basic.robot]: Refined. Thank you!

File: e2e/libs/snapshot/rest.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        parent = self.get(volume_name, parent_id)
+        volume_head = self.get_volume_head(volume_name)
+        if volume_head.name not in parent.children.keys() or volume_head.parent != parent.name:
+            logging(f"Expect snapshot {parent_id} is parent of volume-head, but it's not")
+            time.sleep(self.retry_count)
+
+    def is_existing(self, volume_name, snapshot_id):
+        assert self.get(volume_name, snapshot_id)
+
+    def is_not_existing(self, volume_name, snapshot_id):
```
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/libs/snapshot/rest.py]: Refined. Thank you!

File: e2e/tests/regression/test_basic.robot (hunk 3 above)
Comment (c3y1huang) [reply e2e/tests/regression/test_basic.robot:52]: At the moment, I believe organizing them into folders isn't necessary unless we establish clear groupings for each resource test. Files should suffice for now. What do you think?
```
- e2e/tests/
   +--- volume.robot
   +--- snapshot.robot
   +--- setting.robot
   +--- etc.
```

File: e2e/libs/snapshot/base.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+from abc import ABC, abstractmethod
+from utility.utility import set_annotation
+from utility.utility import get_annotation_value
+
+class Base(ABC):
+
+    ANNOT_ID = "test.longhorn.io/snapshot-id"
```
Comment (c3y1huang) [inline e2e/libs/snapshot/base.py:7]: Should move to the [constant.py](https://github.com/longhorn/longhorn-tests/blob/297533fc8894a203f1777b304c25c2f2119771c6/e2e/libs/utility/constant.py#L7-L8).

File: e2e/libs/volume/base.py [hunk 9]
```diff
@@ -1,8 +1,12 @@
 from abc import ABC, abstractmethod
+from utility.utility import set_annotation
+from utility.utility import get_annotation_value
 
 
 class Base(ABC):
 
+    ANNOT_DATA_CHECKSUM = f'test.longhorn.io/data-checksum-'
```
Comment (c3y1huang) [inline e2e/libs/volume/base.py:8]: Why not reuse [ANNOT_CHECKSUM](https://github.com/longhorn/longhorn-tests/blob/297533fc8894a203f1777b304c25c2f2119771c6/e2e/libs/utility/constant.py#L7)?

**Review by c3y1huang — DISMISSED** [review]

File: e2e/libs/volume/base.py (hunk 9 above)
Comment (yangchiu) [reply e2e/libs/volume/base.py:8]: `ANNOT_CHECKSUM` is the last recorded, but now every checksum of every write has to be recorded.

File: e2e/tests/regression/test_basic.robot (hunk 3 above)
Comment (yangchiu) [reply e2e/tests/regression/test_basic.robot]: Refined. Thank you!

File: e2e/libs/snapshot/base.py (hunk 8 above)
Comment (yangchiu) [reply e2e/libs/snapshot/base.py]: Refined. Thank you!

Comment (c3y1huang) [reply e2e/libs/snapshot/base.py:7]: Looks like this is still here. Will this be moved to a constant file?

File: e2e/libs/volume/base.py (hunk 9 above)
Comment (c3y1huang) [reply e2e/libs/volume/base.py:8]: I see, could you move this to a relevent constant file?

File: e2e/libs/snapshot/base.py (hunk 8 above)
Comment (yangchiu) [reply e2e/libs/snapshot/base.py]: Refined. Thank you!

File: e2e/libs/volume/base.py (hunk 9 above)
Comment (yangchiu) [reply e2e/libs/volume/base.py]: Refined. Thank you!

**Review by chriscchien — APPROVED** [review]
LGTM, went through the code and did not have any concern.

**Pull request merged**